### PR TITLE
Blockade prow plugin (formerly block-paths munger).

### PR DESCRIPTION
**Pull request by cjwagner**
/cc @rmmh

### CODE REVIEW

File: prow/plugins/blockade/blockade.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+/*
+Example config for this plugin:
+
+blockades:
+- repos:
+	- kubernetes-incubator
+	- kubernetes/kubernetes
+	- kubernetes/test-infra
+  blockregexps:
```
Comment (rmmh) [inline prow/plugins/blockade/blockade.go]: document the flow at the top: default allow, deny if a changed file matches any of blockregexps, but allow if it matches any of exception regexps. blocking means adding `do-not-merge/blocked-paths`.

File: prow/plugins/blockade/blockade.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	pluginName        = "blockade"
+	blockedPathsLabel = "do-not-merge/blocked-paths"
```
Comment (rmmh) [inline prow/plugins/blockade/blockade.go]: Should this be part of the config? Bad docs changes had a special label before. Other users might want a different label to be applied.

File: prow/plugins/blockade/blockade.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	return handle(c, pc.PluginConfig.Blockades, &pre)
+}
+
+// blockade is a compiled version of a plugins.Blockade config struct.
+type blockade struct {
+	blockRegexps, exceptionRegexps []*regexp.Regexp
+	explanation                    string
+}
+
+func (bd *blockade) isBlocker(file string) bool {
```
Comment (rmmh) [inline prow/plugins/blockade/blockade.go]: Should be `isBlocked`

File: prow/plugins/blockade/blockade.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	repo := pre.Repo.Name
+	labels, err := c.ghc.GetIssueLabels(org, repo, pre.Number)
+	if err != nil {
+		return err
+	}
+
+	labelPresent := hasBlockedLabel(labels)
+	blockades := compileApplicableBlockades(org, repo, c.log, config)
+	if len(blockades) == 0 && !labelPresent {
+		return nil
```
Comment (rmmh) [inline prow/plugins/blockade/blockade.go]: `// if the label is missing, we assume that we removed any associated comments.`

File: prow/plugins/blockade/blockade.go (hunk 2 above)
Comment (cjwagner) [reply prow/plugins/blockade/blockade.go]: If we let each blockade specify its own label and a blockade is removed, we won't know about that blockade's label anymore and won't be able to remove it from PRs that already have it. 
We could allow partial customization of the label by allowing blockades to specify a label suffix and prefix all labels with `do-not-merge/blockade/` or something like that. Then the plugin can remove any labels with the prefix that no longer apply. That would result in long label names and additional label bloat though.

File: prow/plugins/blockade/blockade.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Example config for this plugin:
```
Comment (spxtr) [inline prow/plugins/blockade/blockade.go]: This comment should be above the actual config definition so that you get it with `go doc`.

**Comment by cjwagner** [comment]
Comments addressed. PTAL

File: prow/plugins/plugins.go [hunk 6]
```diff
@@ -138,6 +138,43 @@ type Configuration struct {
...
+  blockade plugin is defined as a list of these structures. Here is an example of a complete
+  yaml config for the blockade plugin that is composed of 2 Blockade structs:
+
+	blockades:
+	- repos:
+	  - kubernetes-incubator
+	  - kubernetes/kubernetes
+	  - kubernetes/test-infra
+	  blockregexps:
+	  - "docs/.*"
```
Comment (rmmh) [inline prow/plugins/plugins.go]: use ' to quote regexps to avoid interpreting \ escapes

Comment (cjwagner) [reply prow/plugins/plugins.go]: Is that valid yaml? Back ticks don't appear to be a valid replacement for quotes.

Comment (rmmh) [reply prow/plugins/plugins.go]: not back ticks, single quotes

Comment (cjwagner) [reply prow/plugins/plugins.go]: Ah ok. Fixed.

**Pull request merged**